### PR TITLE
S3t1 - Get Single Item (Get All was accomplished originally)

### PR DESCRIPTION
**Pull request by Parsley-Afroman**
I have creates a get single product as we already had a get all products.

Please also take a look at what I have added to the Google Doc to confirm you are happy with what I have added to it.

I haven't sanitised specialChars but I have validated whether they exist and return an alternative string if it does.

### CODE REVIEW

File: Controllers/ProductController.js [hunk 1]
```diff
@@ -6,14 +6,15 @@ const DataCheckers = require("../Services/SanitiseValidateService");
...
-        : JsonResponseService()
-
+    let apiResponse
+    products.length > 0
+        ? apiResponse = JsonResponseService(products, true, 'Success', 200)
+        : apiResponse = JsonResponseService()
     res.json(apiResponse)
 }
 
+
```
Comment (gabrielrowan) [inline Controllers/ProductController.js:16]: I think we should delete 3xline spaces here

File: Services/ProductsService.js [hunk 2]
```diff
@@ -1,14 +1,31 @@
...
     return await connection.query("SELECT `name`, `price`, `SKU` FROM `products`;")
-
 }
 
 const addSingleProduct = async(connection, sku, name, price, stock_level) => {
     return await connection.query('INSERT `products` (`SKU`, `name`, `price`,`stock_level`) VALUES' + `('${sku}', '${name}', '${price}', '${stock_level}');`)
 }
 
+const getSingleProduct = async (connection, SKU) => {
+    const verifiedSKU = async (SKU) => {
```
Comment (gabrielrowan) [inline Services/ProductsService.js:10]: I think we currently have duplicate sanitiseSKU functions - one in this file and one in the sanitiseValidateService. I think we need to keep the sanitise validate logic in the SanitiseValidateService and then call it in the Controller, and keep the Product Service file for the connection.querys only

File: Controllers/ProductController.js [hunk 3]
```diff
@@ -28,6 +29,19 @@ const addSingleProduct = async(req, res) => {
         }
 }
 
+const getSingleProduct = async(req, res) => {
+    const connection = await DbService()
+    const SKU = req.params.SKU
```
Comment (gabrielrowan) [inline Controllers/ProductController.js:34]: I think this line is where we should call the sanitise function, like we did for addSingleProduct. ie: 
const sku = DataCheckers.sanitiseSku(req.body.SKU)

File: Services/ProductsService.js [hunk 4]
```diff
@@ -1,14 +1,31 @@
...
+        const regex = /[A-Za-z]{3}-[0-9A-Za-z]{1,3}-[A-Za-z]{3}/
+        const specChar = /[!@#$%^&*()_+\-=\[\]{};':"\\|,.<>\/?]+/
+        const normaliseString = SKU.toUpperCase()
+        let response
+        if (normaliseString.length > 11 || normaliseString.length < 9 && specChar.test(normaliseString)){
+            response = ''
+        } else {
+            regex.test(SKU)
+                ? response = await connection.query('SELECT `SKU`, `name`, `price`, `stock_level` FROM `products` WHERE `SKU` =' + `'${normaliseString}'`)
```
Comment (gabrielrowan) [inline Services/ProductsService.js:19]: I would add 'LIMIT 1' to this conneciton.query
full sql query would be equivalent of this in SQL 'SELECT `SKU`, `name`, `price`, `stock_level` FROM `products` WHERE `SKU` = 'CRO-3-GRE' LIMIT 1;'
This means that, even when there are duplicates in the database, only 1 item will be returned

**Review by gabrielrowan — CHANGES_REQUESTED** [review]
Hey, looks good. I've added a few comments because I think currently we have 2 sanitiseSKU functions and would be good to have separation of concern in the productservice file and keep the sanitisation in the sanitise file

File: Services/ProductsService.js (hunk 4 above)
Comment (Parsley-Afroman) [reply Services/ProductsService.js:19]: Ye there is no harm in doing that, I merely didn't as there shouldn't be duplicates.....should we be checking for this when adding an item?

File: Controllers/ProductController.js (hunk 1 above)
Comment (Parsley-Afroman) [reply Controllers/ProductController.js:16]: NP

File: Services/ProductsService.js (hunk 2 above)
Comment (Parsley-Afroman) [reply Services/ProductsService.js:10]: Agreed, Amended

File: Controllers/ProductController.js (hunk 3 above)
Comment (Parsley-Afroman) [reply Controllers/ProductController.js:34]: Agreed, Amended

File: Services/ProductsService.js (hunk 4 above)
Comment (Parsley-Afroman) [reply Services/ProductsService.js:19]: Amended

**Pull request merged**